File: dotnet/src/Rerout/Models/Link.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Rerout.Models;

/// <summary>A short link as returned by the Rerout API.</summary>
public sealed record Link
{
    /// <summary>The short link path code.</summary>
    [JsonPropertyName("code")]
    public required string Code { get; init; }

    /// <summary>Fully-qualified short URL — <c>https://{host}/{code}</c>.</summary>
    [JsonPropertyName("short_url")]
    public required string ShortUrl { get; init; }

    /// <summary>Verified custom domain hosting this link, when one is bound.</summary>
    [JsonPropertyName("domain_hostname")]
    public string? DomainHostname { get; init; }

    /// <summary>Destination the redirect resolves to.</summary>
    [JsonPropertyName("target_url")]
    public required string TargetUrl { get; init; }

    /// <summary>Project that owns the link.</summary>
    [JsonPropertyName("project_id")]
    public required string ProjectId { get; init; }

    /// <summary>Unix seconds — expiration. <c>null</c> for permanent links.</summary>
    [JsonPropertyName("expires_at")]
    public long? ExpiresAt { get; init; }

    /// <summary>Whether the link is currently active.</summary>
    [JsonPropertyName("is_active")]
    public required bool IsActive { get; init; }

    /// <summary>Override social preview title.</summary>
    [JsonPropertyName("seo_title")]
    public string? SeoTitle { get; init; }

    /// <summary>Override social preview description.</summary>
    [JsonPropertyName("seo_description")]
    public string? SeoDescription { get; init; }

    /// <summary>Override social preview image URL.</summary>
    [JsonPropertyName("seo_image_url")]
    public string? SeoImageUrl { get; init; }

    /// <summary>Override preview canonical URL.</summary>
    [JsonPropertyName("seo_canonical_url")]
    public string? SeoCanonicalUrl { get; init; }

    /// <summary>Whether the preview HTML should be indexed.</summary>
    [JsonPropertyName("seo_noindex")]
    public bool SeoNoindex { get; init; }

    /// <summary>Unix seconds — last SEO field change.</summary>
    [JsonPropertyName("seo_updated_at")]
    public long? SeoUpdatedAt { get; init; }

    /// <summary>Unix seconds — link creation time.</summary>
    [JsonPropertyName("created_at")]
    public required long CreatedAt { get; init; }

    /// <summary>Unix seconds — last mutation.</summary>
    [JsonPropertyName("updated_at")]
    public required long UpdatedAt { get; init; }

    /// <summary>Read-only tags attached to the link. Empty when none are set.</summary>
    [JsonPropertyName("tags")]
    public IReadOnlyList<Tag> Tags { get; init; } = [];

    /// <summary>Smart Links — whether a password is required to follow this link.</summary>
    [JsonPropertyName("password_protected")]
    public bool PasswordProtected { get; init; }

    /// <summary>Smart Links — click cap, or <c>null</c> when uncapped.</summary>
    [JsonPropertyName("max_clicks")]
    public long? MaxClicks { get; init; }

    /// <summary>Smart Links — total clicks recorded against this link.</summary>
    [JsonPropertyName("click_count")]
    public long ClickCount { get; init; }

    /// <summary>Smart Links — whether conversion tracking is enabled.</summary>
    [JsonPropertyName("track_conversions")]
    public bool TrackConversions { get; init; }

    /// <summary>Smart Links — ordered geo/device routing rules.</summary>
    [JsonPropertyName("routing_rules")]
    public IReadOnlyList<RoutingRule> RoutingRules { get; init; } = [];

    /// <summary>Smart Links — weighted A/B destinations.</summary>
    [JsonPropertyName("ab_variants")]
    public IReadOnlyList<AbVariant> AbVariants { get; init; } = [];
}

/// <summary>
/// A geo/device routing rule (Smart Links). When the condition matches the
/// incoming request, the redirect resolves to <see cref="TargetUrl"/> instead of
/// the link's default destination.
/// </summary>
/// <remarks>
/// <see cref="ConditionType"/> is <c>country</c> or <c>device</c>;
/// <see cref="ConditionOp"/> is <c>is</c>, <c>is_not</c>, or <c>in</c>. These are
/// sent verbatim and not validated client-side.
/// </remarks>
public sealed record RoutingRule
{
    /// <summary>What to match against — <c>country</c> or <c>device</c>.</summary>
    [JsonPropertyName("condition_type")]
    public required string ConditionType { get; init; }

    /// <summary>Comparison operator — <c>is</c>, <c>is_not</c>, or <c>in</c>.</summary>
    [JsonPropertyName("condition_op")]
    public required string ConditionOp { get; init; }

    /// <summary>Value(s) to compare against (e.g. <c>ZA</c>, <c>US,GB</c>, <c>mobile</c>).</summary>
    [JsonPropertyName("condition_value")]
    public required string ConditionValue { get; init; }

    /// <summary>Destination when the rule matches.</summary>
    [JsonPropertyName("target_url")]
    public required string TargetUrl { get; init; }
}

/// <summary>
/// A weighted A/B destination (Smart Links), as returned by the API. To supply
/// variants on create/update, use <see cref="AbVariantInput"/>.
/// </summary>
public sealed record AbVariant
{
    /// <summary>Stable variant id assigned by the server.</summary>
    [JsonPropertyName("id")]
    public required long Id { get; init; }

    /// <summary>Destination for this variant.</summary>
    [JsonPropertyName("target_url")]
    public required string TargetUrl { get; init; }

    /// <summary>Relative weight in the split.</summary>
    [JsonPropertyName("weight")]
    public required int Weight { get; init; }
}

/// <summary>
/// A weighted A/B destination as supplied on create/update. <see cref="Weight"/>
/// is optional and defaults server-side when omitted.
/// </summary>
public sealed record AbVariantInput
{
    /// <summary>Destination for this variant.</summary>
    [JsonPropertyName("target_url")]
    public required string TargetUrl { get; init; }

    /// <summary>Relative weight in the split. Omit to default server-side.</summary>
    [JsonPropertyName("weight")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Weight { get; init; }
}

File: php/src/Models/Link.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * A short link as returned by the Rerout API.
 *
 * Fields mirror the server-side `LinkResponse` shape one-to-one so JSON is
 * parsed without transformation.
 */
final readonly class Link
{
    /**
     * @param string      $code            Short link path code.
     * @param string      $shortUrl        Fully-qualified short URL — `https://{host}/{code}`.
     * @param string|null $domainHostname  Verified custom domain hosting this link, when bound.
     * @param string      $targetUrl       Destination the redirect resolves to.
     * @param string      $projectId       Project that owns the link.
     * @param int|null    $expiresAt       Unix seconds — expiration. Null for permanent links.
     * @param bool        $isActive        Whether the link is currently active.
     * @param string|null $seoTitle        Override social preview title.
     * @param string|null $seoDescription  Override social preview description.
     * @param string|null $seoImageUrl     Override social preview image URL.
     * @param string|null $seoCanonicalUrl Override preview canonical URL.
     * @param bool        $seoNoindex      Whether the preview HTML should be indexed.
     * @param int|null    $seoUpdatedAt    Unix seconds — last SEO field change.
     * @param int         $createdAt       Unix seconds — link creation time.
     * @param int                $updatedAt         Unix seconds — last mutation.
     * @param list<Tag>          $tags              Tags attached to the link. Read-only;
     *                                              empty when none are bound.
     * @param bool               $passwordProtected Whether the link requires a password.
     * @param int|null           $maxClicks         Click cap before the link stops resolving. Null for unlimited.
     * @param int                $clickCount        Total clicks recorded so far.
     * @param bool               $trackConversions  Whether conversion tracking is enabled.
     * @param list<RoutingRule>  $routingRules      Smart-routing rules evaluated in order.
     * @param list<AbVariant>    $abVariants        A/B test variants traffic is split across.
     */
    public function __construct(
        public string $code,
        public string $shortUrl,
        public ?string $domainHostname,
        public string $targetUrl,
        public string $projectId,
        public ?int $expiresAt,
        public bool $isActive,
        public ?string $seoTitle,
        public ?string $seoDescription,
        public ?string $seoImageUrl,
        public ?string $seoCanonicalUrl,
        public bool $seoNoindex,
        public ?int $seoUpdatedAt,
        public int $createdAt,
        public int $updatedAt,
        public array $tags = [],
        public bool $passwordProtected = false,
        public ?int $maxClicks = null,
        public int $clickCount = 0,
        public bool $trackConversions = false,
        public array $routingRules = [],
        public array $abVariants = [],
    ) {
    }

    /**
     * Parse a {@see Link} from an API JSON envelope.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            code: self::requireString($data, 'code'),
            shortUrl: self::requireString($data, 'short_url'),
            domainHostname: self::nullableString($data, 'domain_hostname'),
            targetUrl: self::requireString($data, 'target_url'),
            projectId: self::requireString($data, 'project_id'),
            expiresAt: self::nullableInt($data, 'expires_at'),
            isActive: self::requireBool($data, 'is_active'),
            seoTitle: self::nullableString($data, 'seo_title'),
            seoDescription: self::nullableString($data, 'seo_description'),
            seoImageUrl: self::nullableString($data, 'seo_image_url'),
            seoCanonicalUrl: self::nullableString($data, 'seo_canonical_url'),
            seoNoindex: self::optionalBool($data, 'seo_noindex', true),
            seoUpdatedAt: self::nullableInt($data, 'seo_updated_at'),
            createdAt: self::requireInt($data, 'created_at'),
            updatedAt: self::requireInt($data, 'updated_at'),
            tags: self::mapTags($data['tags'] ?? []),
            passwordProtected: self::optionalBool($data, 'password_protected', false),
            maxClicks: self::nullableInt($data, 'max_clicks'),
            clickCount: self::optionalInt($data, 'click_count', 0),
            trackConversions: self::optionalBool($data, 'track_conversions', false),
            routingRules: self::mapRoutingRules($data['routing_rules'] ?? []),
            abVariants: self::mapAbVariants($data['ab_variants'] ?? []),
        );
    }

    /**
     * @return list<Tag>
     */
    private static function mapTags(mixed $raw): array
    {
        if (!is_array($raw)) {
            return [];
        }
        $out = [];
        foreach ($raw as $row) {
            if (is_array($row)) {
                /** @var array<string, mixed> $row */
                $out[] = Tag::fromArray($row);
            }
        }

        return $out;
    }

    /**
     * @return list<RoutingRule>
     */
    private static function mapRoutingRules(mixed $raw): array
    {
        if (!is_array($raw)) {
            return [];
        }
        $out = [];
        foreach ($raw as $row) {
            if (is_array($row)) {
                /** @var array<string, mixed> $row */
                $out[] = RoutingRule::fromArray($row);
            }
        }

        return $out;
    }

    /**
     * @return list<AbVariant>
     */
    private static function mapAbVariants(mixed $raw): array
    {
        if (!is_array($raw)) {
            return [];
        }
        $out = [];
        foreach ($raw as $row) {
            if (is_array($row)) {
                /** @var array<string, mixed> $row */
                $out[] = AbVariant::fromArray($row);
            }
        }

        return $out;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireString(array $data, string $key): string
    {
        $value = $data[$key] ?? null;
        if (!is_string($value)) {
            throw new InvalidArgumentException("Expected string field '{$key}' in Link payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function nullableString(array $data, string $key): ?string
    {
        $value = $data[$key] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_string($value)) {
            throw new InvalidArgumentException("Expected nullable string field '{$key}' in Link payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireInt(array $data, string $key): int
    {
        $value = $data[$key] ?? null;
        if (!is_int($value)) {
            throw new InvalidArgumentException("Expected int field '{$key}' in Link payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function nullableInt(array $data, string $key): ?int
    {
        $value = $data[$key] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_int($value)) {
            throw new InvalidArgumentException("Expected nullable int field '{$key}' in Link payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireBool(array $data, string $key): bool
    {
        $value = $data[$key] ?? null;
        if (!is_bool($value)) {
            throw new InvalidArgumentException("Expected bool field '{$key}' in Link payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function optionalBool(array $data, string $key, bool $default): bool
    {
        $value = $data[$key] ?? null;
        if ($value === null) {
            return $default;
        }
        if (!is_bool($value)) {
            throw new InvalidArgumentException("Expected bool field '{$key}' in Link payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function optionalInt(array $data, string $key, int $default): int
    {
        $value = $data[$key] ?? null;
        if ($value === null) {
            return $default;
        }
        if (!is_int($value)) {
            throw new InvalidArgumentException("Expected int field '{$key}' in Link payload.");
        }

        return $value;
    }
}
